WP Super Cache: preload notification shows last 5 preloaded files now

donnchawp opened the pull request:
When the plugin is preloading pages it displays the last preloaded URL. This patch changes that so it show the last 5 preloaded URLs, with the newest URL coming up at the bottom, and the oldest being removed off the top.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion


## Does this pull request change what data or activity we track or use?
No

## Testing instructions:
Apply patch
Run a preload and  ten seconds or so after it starts,  you will see the last 5 URLs that have been preloaded appear in the admin notice area of the preload page.

![BEFORE](https://github.com/Automattic/jetpack/assets/5656673/f001ea6e-060c-4633-b61a-057f2a7fac7e)
Before


![AFTER](https://github.com/Automattic/jetpack/assets/5656673/a99de418-419a-4195-b904-176c326dfb98)
After

donnchawp committed:
Make the preload notification into a list of the last 5 urls

donnchawp committed:
changelog

donnchawp committed:
3 second gap, and fix stopping.

Repeats have to go for 30 rather than 10, just in case a page is slow
loading.

thingalon commented:
Testing this out, I had a bit of an odd experience:
https://github.com/Automattic/jetpack/assets/1369626/2ad2ac22-3c40-4678-93eb-e0ef78af1fd1

I clicked preload now, and waited for more than 10 seconds. Nothing appeared to happen. So I refreshed the page, and nothing was happening. Tried again, saw the preload in progress, where it reported the same page twice. Nothing further seemed to happen, so I refreshed again to what appeared to be a completed process.

Was I clicking around too impatiently?

donnchawp commented:
There used to be a notification that preload was going to run in X seconds when you reloaded the page but for some reason it's not working now. So, what you saw is expected, apart from the URL showing up twice, but that might be because you have only one or not many posts in your test site? The JS notification script loads the post URL every 3 seconds. It is supposed to check for repeats though, so it shouldn't have shown the URL twice.
The notification system doesn't actually show a "Preload finished" message. For most sites, with hundreds or thousands of posts, the site owner isn't going to see it anyway.

donnchawp committed:
Updates to "refresh notification" on preload page.

* Only show the "Full refresh" message.
* Show it regardless of time to next preload interval. Also when that
  interval was 0 it wouldn't show at all.

donnchawp committed:
Remove the "It will begin caching" message as notification fixed

donnchawp commented:
@thingalon - I fixed the notification. If the "preload interval" was set to 0 it wouldn't show at all, so I just removed that check.
It used to show a notification on the "next preload" which was a 30 second pause, but as it's only 3 seconds now, I removed it. It only shows a notification on the "Full preload" at the start of a preload.

Commit:
Store extra info about current preload for a more informative ui

thingalon commented:
@donnchawp I've tinkered with making the preload notification more informative using a bit more JavaScript, and storing more state in the preload info txt file. It also uses wp-ajax to fetch the preload status, ensuring that the preload will tick forward even if you are the only visitor on your site.

Please test it and review it - ed9458015f I am worried about breaking something with the change :)

dilirity commented on projects/plugins/super-cache/inc/preload-notification.js:
```diff
@@ -1,18 +1,92 @@
+
+	update_preload_status( wpsc_preload_ajax.preload_status );
+
+	setInterval( () => {
+		jQuery.post(
+			wpsc_preload_ajax.ajax_url,
+			{
+				action: 'wpsc_get_preload_status',
+			},
+			( json ) => {
+				if ( ! json || ! json.success ) {
+					return;
+				}
+
+				update_preload_status( json.data );
+			}
+		);
+	}, CHECK_INTERVAL );
```
I'm not a huge fan of relying on the user's connection being fast enough to finish the request in the time specified in `CHECK_INTERVAL`. I know it's an edge case, but it makes my senses tingle :D

donnchawp replied:
It's 3 seconds, and it used to be 1 second. I don't think it's much to worry about though as it's only for the notification. The AJAX request is fairly light weight too.

thingalon replied:
🤔 Defensive code is a good idea. In 4d41ccdc4f I made it use `setTimeout` which is called after each request completes so each will be 3 seconds after the last is completed.

It will also stop if one of the update requests fails, so that we don't keep polling a broken server. If users need to resume after an error they can refresh the page.

dilirity commented on projects/plugins/super-cache/inc/preload-notification.js:
```diff
@@ -1,18 +1,92 @@
+
+	update_preload_status( wpsc_preload_ajax.preload_status );
+
+	setInterval( () => {
+		jQuery.post(
+			wpsc_preload_ajax.ajax_url,
+			{
+				action: 'wpsc_get_preload_status',
+			},
+			( json ) => {
+				if ( ! json || ! json.success ) {
+					return;
+				}
+
+				update_preload_status( json.data );
+			}
+		);
+	}, CHECK_INTERVAL );
```
The request continues even after the preload has finished.

donnchawp replied:
I had a fix for that but it was clobbered. The fix was to check for duplicates, but the code @thingalon added has a "running" status, which I could use to stop it.

donnchawp replied:
Unfortunately, "running" is set to false when the preload is scheduled to start, so I think we'll have to go back to checking for duplicate URLs again.
Once #31022 is merged I'll use wpsc_is_preload_active() to stop the JS files being enqueued, or do we mind it doing an AJAX request every 3 seconds on the preload settings page?

thingalon replied:
In 406ea75214 , I updated the preload code to vary the refresh speed based on whether preload is running (or is scheduled to run soon). If it's running, it refreshes every 3 seconds. If it's idle or the next preload is more than a minute away, it updates every 30 seconds.

dilirity commented on projects/plugins/super-cache/inc/preload-notification.js:
```diff
@@ -1,18 +1,92 @@
+			if ( data.previous ) {
+				const p = jQuery( '<p>' );
+				p.append( jQuery( '<b>' ).text( __( 'Last preload finished:', 'wp-super-cache' ) ) );
+				p.append( jQuery( '<span>' ).text( new Date( data.previous * 1000 ).toLocaleString() ) );
```
Missing space after the colon :)
<img width="294" alt="CleanShot 2023-06-05 at 18 45 43@2x" src="https://github.com/Automattic/jetpack/assets/11799079/0a0e486b-ed95-40a3-979a-45749b6a094b">

thingalon replied:
Fixed in 232a35aff5

dilirity commented on projects/plugins/super-cache/inc/preload-notification.js:
```diff
@@ -1,18 +1,92 @@
+		if ( data.running ) {
+			const panel = jQuery( '<div class="notice notice-warning">' );
+
+			panel.append( jQuery( '<b>' ).text( __( 'Preloading', 'wp-super-cache' ) ) );
```
Missing paragraph around the `b` tag, causes it's spacing to look off
<img width="589" alt="image" src="https://github.com/Automattic/jetpack/assets/11799079/a32b414a-cfbc-41f9-bcd0-ac54e8065697">

Maybe change it to an `H4`?

thingalon replied:
Fixed in d58e6b8d1b. Weirdly, h4 was too far to the left too - so I just kept it simple and added the missing p

dilirity commented on projects/plugins/super-cache/inc/preload-notification.js:
```diff
@@ -1,18 +1,92 @@
+			const panel = jQuery( '<div class="notice notice-info">' );
+
+			if ( data.next ) {
+				const diff = data.next - Math.floor( Date.now() / 1000 );
+				const seconds = diff % 60;
+				const minutes = Math.floor( diff / 60 ) % 60;
+				const hours = Math.floor( diff / 3600 ) % 24;
```
I got a weird result while waiting for the preload to start
<img width="443" alt="image" src="https://github.com/Automattic/jetpack/assets/11799079/352922f4-ff79-4e7a-a1f1-249f46b9a4b9">

thingalon replied:
I can't replicate this, but I guess it might happen if the next run time is in the past.

So I've added a guard against that in d2dfe131bb, which guarantees the diff between now and the next-run time never dips below 0.

dilirity commented on projects/plugins/super-cache/inc/preload-notification.js:
Can the JS file be moved to the root `js` directory? Not a big thing, just wanted to avoid mixing file types.

I just noticed that there's another JS file in there.

donnchawp replied:
Yes, good idea! The inc directory predates the js directory.

thingalon replied:
Done in 4447abfc82.

dilirity reviewed (changes requested):
Good ol' jQuery. I'm not sure who to send the feedback to :D

cc @thingalon just in case.

I really like the new UI! No more manually refreshing the page - love it.

GitHub Actions was kind enough to remind us that we can remove one of the files from the linter exclude lists:
<img width="825" alt="image" src="https://github.com/Automattic/jetpack/assets/11799079/d8eb0567-a4be-498c-a1fb-bc909704e618">

I've also left some comments.

donnchawp commented:
The "Cancel Preload" button reverts back to "Preload Cache Now". I'm not sure yet what is causing that to happen, but I want to merge #31022 first which will cause quite a few conflicts with this patch.

donnchawp committed:
Merge remote-tracking branch 'origin/trunk' into update/preload_notification

Commit:
Move preload js into js dir

Commit:
Use setTimeout instead of setInterval, so the next time is based on the end of the previous request

Commit:
clear the preload data when stopping

Commit:
Update preload every 3 seconds when something is happening, every 30 seconds otherwise

Commit:
<p> wrap around preloading title

Commit:
Paranoid math guard against negative time in the future

Commit:
space

Commit:
cleaned up file

thingalon commented:
@dilirity thank you for the thorough review. I believe I've addressed all of your points. Please check it out :-)

thingalon commented:
> The "Cancel Preload" button reverts back to "Preload Cache Now". I'm not sure yet what is causing that to happen, but I want to merge #31022 first which will cause quite a few conflicts with this patch.

It seems to be working to me. How can I reproduce the issue?

donnchawp committed:
Check the counter and transient for "preload" button status.

* Will replace those extra checks with wpsc_is_preload_active later

donnchawp commented:
The issue was, $currently_preloaded wasn't set to true because [this line](https://github.com/Automattic/jetpack/pull/31017/files#diff-49e24876fd767aa7d12d7018a180f19e42dc62b47484a137fe4d42fdb31d10f6L1018) was removed. I just added that check in partials/preload.php and it appears to be working again,
However, the "Preloading is currently running" message is still showing, after I cancelled the preload so I need to clean that up.

donnchawp committed:
Set the idle status when cancelling during taxonomy preload

donnchawp commented:
The issue there was the idle status was't being set if preload was cancelled during taxonomy preload. Fixed that in https://github.com/Automattic/jetpack/pull/31017/commits/fd89b9c119ccaf89422137a395d90c3bd04daf20 :)

dilirity reviewed (approved):
Thanks guys! This is looking great.

:shipit:
